File: DataStructures/LinkedLists/LinkedList.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

/* Node definition */
typedef struct Node 
{
    int data;
    struct Node *next;
}
Node;

/* Initialize a NULL head Node */
Node *head = NULL;

/* Define our functions here */
void add_front(int data); // O(1)
int remove_front(); // O(1)
void add_back(int data); // O(N)
int remove_end(); // O(N)
bool search(int data); // O(N)
void show_items();

/* Test Client */
int main(void)
{
    // Add some nodes to front
    add_front(100);
    add_front(200);
    add_front(300);
    add_back(400);
    add_back(500);
    remove_front();
    remove_end();
    printf("200 in list: %s\n", search(200) ? "true" : "false");
    printf("600 in list: %s\n", search(600) ? "true" : "false");
    // Print the nodes
    show_items();
}

/* Functions */
void add_front(int data)
{
    // Create a new Node and add to front
    Node *n = malloc(sizeof(Node));
    n->data = data;
    n->next = head;
    head = n;  
}

void add_back(int data)
{
    if (head == NULL) // Check if list is empty
    {
        add_front(data);
    }
    else // Iterate until we reach end of list and insert Node after last Node
    {
        Node *new_node = malloc(sizeof(Node)); // Create new Node
        new_node->data = data;
        new_node->next = NULL;
        Node *t = head;
        while (t->next != NULL)
        {
            t = t->next;
        }
        t->next = new_node;
    }
}

int remove_front()
{
    Node *t = head;
    int data = head->data;
    head = head->next;
    free(t);
    return data;
}

int remove_end()
{
    Node *n = head;
    while (n->next->next != NULL)
    {
        n = n->next;
    }
    int data = n->next->data;
    n->next = NULL;
    free(n->next);
    return data;
}

bool search(int data)
{
    Node *n = head;
    while (n != NULL)
    {
        if (n->data == data)
        {
            return true;
        }
        n = n->next;
    }
    return false;
}

void show_items()
{
    Node *n = head;
    printf("[Head:");
    while (n != NULL)
    {
        if (n->next == NULL)
        {
            printf("%i", n->data);
        }
        else
        {
            printf("%i->", n->data);
        }
        n = n->next;
    }
    printf("]\n");
}

File: DataStructures/LinkedLists/LinkedList.py
class Node:

    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:

    def __init__(self):
        self.head = None
        self.size = 0

    def get_size(self):
        return self.size

    # O(1) operation
    def insert_start(self, data):
        self.size += 1
        if self.head is None:
            self.head = Node(data)
        else:
            new_node = Node(data)
            new_node.next = self.head
            self.head = new_node

    # O(1) operation
    def remove_start(self):
        if self.head is None:
            return None
        self.size -= 1
        data = self.head.data
        self.head = self.head.next
        return data

    # O(N) operation
    def insert_end(self, data):
        if self.head is None:
            self.insert_start(data)
        else:
            self.size += 1
            n = self.head
            while n.next is not None:
                n = n.next
            n.next = Node(data)

    # O(N) operation
    def remove_end(self):
        if self.head is None:
            return
        elif self.size == 1:
            self.remove_start()
        else:
            self.size -= 1
            n = self.head
            prev = None
            while n.next is not None:
                prev = n
                n = n.next
            data = n.data
            prev.next = None
            return data

    def traverse(self):
        n = self.head
        while n is not None:
            print(n.data)
            n = n.next


def main():

    ll = LinkedList()
    ll.insert_start(300)
    ll.insert_start(200)
    ll.insert_start(100)
    ll.insert_end(400)
    ll.insert_end(500)
    ll.insert_end(600)
    ll.traverse()
    print('-------------')
    ll.remove_end()
    ll.remove_end()
    ll.traverse()


main()